### ISSUE TITLE
Invoke-Pester should have a summary at the end

### ISSUE THREAD

**Issue by DarkLite1** [pester/Pester#1886]
## Summary of the feature request
It would be to have a summary on file level when the `Invoke-Pester` command is finished. At the moment, when running `Invoke-Pester` on multiple files all executed test scripts are colored green. This makes it hard to quickly see which script failed by just looing at the console. 

![image](https://user-images.githubusercontent.com/10828570/112635020-95a89480-8e3b-11eb-8d65-f78dc09ddf10.png)

For the example above we first issued `$WarningPreference = 'SilentlyContinue'` to suppress all warnings, otherwise it would really be a hunt through lines of warnings in the console to find the error line to see which test script failed.

## How should it work? 

It would really be cool to have a small summary/overview at the end. in the console Something like this:

```
File                  | Passed | Failed | Skipped
[+] Script1.Tests.ps1 |   15      |     0     |      0     # green
[+] Script2.Tests.ps1 |   70      |     1     |      0     # red
[+] Script3.Tests.ps1 |   18      |     0     |      0     # green
````

**Comment by fflaten** [comment]
This is easily achieved using the resultobject returned from `Invoke-Pester -Passthru`. Then you can format it however you'd like 🙂 

```
$p = Invoke-Pester -Path ./Samples/demoIssue1234.tests.ps1, ./Samples/demoIssue1853.tests.ps1, ./Samples/demoBeforeDiscovery.tests.ps1 -PassThru
$p.Containers | Format-Table Item, PassedCount, FailedCount, SkippedCount                                                                                 


Item                                                     PassedCount FailedCount SkippedCount
----                                                     ----------- ----------- ------------
/workspaces/Pester/Samples/demoIssue1234.tests.ps1                 0           1            0
/workspaces/Pester/Samples/demoIssue1853.tests.ps1                 0           1            0
/workspaces/Pester/Samples/demoBeforeDiscovery.tests.ps1          15           0            5
```

**Comment by DarkLite1** [comment]
Thank you for the tip :) That is indeed a nice workaround. But it would be great to have Pester do that by default. A test suite like Jest for example does something similar.

**Comment by fflaten** [comment]
A summary like that can be very user-specific. The fact that you can interact with a result-object without export/import XML/JSON is a great benefit with PowerShell. It allows you to quickly adapt based on your specific requirements. A personal helper function in your profile or project would also make it easy to run often.

```
# StartLine = Location of It-block
$p.Tests | ft ExpandedPath, Result, @{n="FilePath";e={ $_.Block.BlockContainer.Item }}, StartLine   


ExpandedPath                                                  Result FilePath                                           StartLine
------------                                                  ------ --------                                           ---------
Summary Test.Root test                                        Passed /workspaces/Pester/Samples/demoIssue1886.tests.ps1         3
Summary Test.Context 1.Test 1                                 Failed /workspaces/Pester/Samples/demoIssue1886.tests.ps1         8
Summary Test.Context 1.Inner Context 1.Test 1                 Failed /workspaces/Pester/Samples/demoIssue1886.tests.ps1        13
Summary Test.Inner Describe 1.Test 2                          Failed /workspaces/Pester/Samples/demoIssue1886.tests.ps1        20
Summary Test.Inner Describe 1.Inner Describe Context 2.Test 2 Failed /workspaces/Pester/Samples/demoIssue1886.tests.ps1        25
Summary Test 2.Root test                                      Passed /workspaces/Pester/Samples/demoIssue1886.tests.ps1        34
```

or for failed tests only:

```
$p.Failed | ft ExpandedPath, @{n="ErrorMessage";e={ $_.ErrorRecord.Exception.Message }}, @{n="FilePath";e={ $_.Block.BlockContainer.Item }}, StartLine                 


ExpandedPath                                                  ErrorMessage           FilePath                                           StartLine
------------                                                  ------------           --------                                           ---------
Summary Test.Context 1.Test 1                                 Expected 2, but got 1. /workspaces/Pester/Samples/demoIssue1886.tests.ps1         8
Summary Test.Context 1.Inner Context 1.Test 1                 Expected 2, but got 1. /workspaces/Pester/Samples/demoIssue1886.tests.ps1        13
Summary Test.Inner Describe 1.Test 2                          Expected 2, but got 1. /workspaces/Pester/Samples/demoIssue1886.tests.ps1        20
Summary Test.Inner Describe 1.Inner Describe Context 2.Test 2 Expected 2, but got 1. /workspaces/Pester/Samples/demoIssue1886.tests.ps1        25
```

Just to be clear, I'm not declining your feature request so lets see what others think 🙂 

Apart from the current `-Output` options, my suggestion would be to provide more examples on https://pester.dev on how to generate reports like above using the output of `-Passthru`.

**Comment by DarkLite1** [comment]
I hear you and I definitely get that this is easy for a user to build. However, given that Pester already does some output in colors for the Discovery phase and also in colors for the actual run, my feeling is that there should be a default summary at the end. This can of course be hidden with a switch statement like say:

```powershell
Invoke-Pester -NoSummary
```

However, if this will be rejected the least I think that should be corrected is to no longer display the line in green with a plus `[+]` sign in front of script name but have it red with a negative sign `[-]` when a test for that file fails. That is currently not the case. If one test in the file fails, the ling is still green with a plus.

**Comment by nohwnd** [comment]
@DarkLite1 it is not printing green message for the file, your Permission audit.Tests.ps1 passed, but Permission matrix.Tests.ps1 failed. 

Here I broke one test in Mocks.Tests.ps1, it does not report Mock.Tests.ps1 as successful.
![image](https://user-images.githubusercontent.com/5735905/112766867-ff7b9680-9013-11eb-8542-2b0f17ae60d9.png)

I have a summary of failed files in my P module and I can say it's mostly useless, unless you want to open that file and run all tests in it again to see what failed. On the other hand the Pester output only shows you failed tests by default and even formats the callstack in a way that you can click it, and go directly to the failing line if you are running in VSCode.

As @fflaten said this is very user specific. And personally I find the output to be good balance of information and noise, with passed files taking very little space on screen and failed tests showing enough info. But it is of course personal preference, and if you have a ton of Write-Warning among all of this it can get overwhelming. I have the same in my P tests where mix of pester output and the other test framework is mixed and finding the failed test is difficult. But it also does not have the result object, otherwise I would use it in the way described above.

**Comment by DarkLite1** [comment]
Thank you for getting back to me Jakub. You are right, when I look at this with a fresh head now I can see that the file `Permission Matrix.tests` failed. I think I just got confused with the other green ones that were displaying the file name and for the failed file only the error is displayed of the test and not the file name. I like your work on the clickable link in vs code, although I haven't used it yet.

So yes, this must be a personal preference thing. Closing this one as I'm the only one supporting this idea :P Thanks again for the already awesome module. Really appreciate the work that went into this.
